### Part 1: Manipulating Data with Pandas library

In [ ]:
import pandas as pd
import numpy as np

In [ ]:
movie_df = pd.read_csv("movies_metadata.csv")

/usr/local/lib/python3.7/dist-packages/IPython/core/interactiveshell.py:3326: DtypeWarning: Columns (10) have mixed types.Specify dtype option on import or set low_memory=False.
  exec(code_obj, self.user_global_ns, self.user_ns)


## Create a small DataFrame small_df that only has the following features: title, release_date, budget, revenue, runtime, and genres.



In [ ]:
small_df = movie_df.loc[:, ["title", "release_date", "budget", "revenue", "runtime", "genres"]]
small_df

title release_date    budget      revenue  \
0                        Toy Story   1995-10-30  30000000  373554033.0   
1                          Jumanji   1995-12-15  65000000  262797249.0   
2                 Grumpier Old Men   1995-12-22         0          0.0   
3                Waiting to Exhale   1995-12-22  16000000   81452156.0   
4      Father of the Bride Part II   1995-02-10         0   76578911.0   
...                            ...          ...       ...          ...   
45461                       Subdue          NaN         0          0.0   
45462          Century of Birthing   2011-11-17         0          0.0   
45463                     Betrayal   2003-08-01         0          0.0   
45464             Satan Triumphant   1917-10-21         0          0.0   
45465                     Queerama   2017-06-09         0          0.0   

       runtime                                             genres  
0         81.0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...  
1        104.0  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...  
2        101.0  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...  
3        127.0  [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...  
4        106.0                     [{'id': 35, 'name': 'Comedy'}]  
...        ...                                                ...  
45461     90.0  [{'id': 18, 'name': 'Drama'}, {'id': 10751, 'n...  
45462    360.0                      [{'id': 18, 'name': 'Drama'}]  
45463     90.0  [{'id': 28, 'name': 'Action'}, {'id': 18, 'nam...  
45464     87.0                                                 []  
45465     75.0                                                 []  

[45466 rows x 6 columns]

## Check out the data types of the different features. What is the type of “budget”?

In [ ]:
small_df.dtypes

title            object
release_date     object
budget           object
revenue         float64
runtime         float64
genres           object
dtype: object

In [ ]:
small_df["budget"].dtypes

dtype('O')

In [ ]:
df = small_df["budget"]

## Convert the budget feature to float. What is the result? 

In [ ]:
small_df["budget"] = small_df["budget"].astype(float)
small_df["budget"]

ValueError: ignored

## Create a function to_float(number) to convert every number field in budget to float and, if that fails, convert it to NaN. 

In [ ]:
import numpy as np
import sys

def to_float(number):
  try:
    return float(number)
  except:
    return(np.nan)

# Apply the to_float function to all values in the budget column, then try converting to float using pandas astype. Check out your result by getting the data types for all features. 

In [ ]:
small_df["budget"] = small_df["budget"].apply(to_float)


In [ ]:
small_df["budget"].astype('float')

0        30000000.0
1        65000000.0
2               0.0
3        16000000.0
4               0.0
            ...    
45461           0.0
45462           0.0
45463           0.0
45464           0.0
45465           0.0
Name: budget, Length: 45466, dtype: float64

# Define a new feature called “year” that represents the year of release by using the datetime functionality from pandas as following:
o Convert release_date into pandas datetime format.

o Extract year from the datetime.


In [ ]:
small_df["release_date"] = pd.to_datetime(small_df["release_date"], errors='coerce')
small_df["release_date"]

0       1995-10-30
1       1995-12-15
2       1995-12-22
3       1995-12-22
4       1995-02-10
           ...    
45461          NaT
45462   2011-11-17
45463   2003-08-01
45464   1917-10-21
45465   2017-06-09
Name: release_date, Length: 45466, dtype: datetime64[ns]

In [ ]:
small_df["year"] = pd.DatetimeIndex(small_df["release_date"]).year

In [ ]:
small_df["year"]

0        1995.0
1        1995.0
2        1995.0
3        1995.0
4        1995.0
          ...  
45461       NaN
45462    2011.0
45463    2003.0
45464    1917.0
45465    2017.0
Name: year, Length: 45466, dtype: float64

## What are the oldest movies available in this dataset?

In [ ]:
#To find the oldest movie
#oldest_date = small_df["release_date"].min()
#print("Oldest movie is from", oldest_date.strftime('%d-%m-%Y'))
#oldest_movie = small_df[small_df["release_date"] == oldest_date]
#print("The oldest movie is ", oldest_movie["title"].to_string(index=False))

#To find the 10 oldest movies
old_movies = small_df.sort_values('release_date').head(10)
print("The name of the 10 old movies are:\n",old_movies[["title", "year"]].to_string(index=False))

The name of the 10 old movies are:
                         title   year
             Passage of Venus 1874.0
   Sallie Gardner at a Gallop 1878.0
              Buffalo Running 1883.0
  Man Walking Around a Corner 1887.0
             Accordion Player 1888.0
Traffic Crossing Leeds Bridge 1888.0
    London's Trafalgar Square 1890.0
                    Mosquinha 1890.0
          Monkeyshines, No. 1 1890.0
          Monkeyshines, No. 2 1890.0


# What are the most successful (revenue feature) movies of all time?

In [ ]:
#To find the most successful movie
#rev_max = small_df["revenue"].max()
#print("The highest revenue is", rev_max)
#rev_max_name = list(small_df[small_df["revenue"] == rev_max]["title"])
#print("The oldest movie is ", rev_max_name[0])

#To find the most successful movies

high_movies = small_df.sort_values('revenue', ascending=False).head(10)
print("10 most successful movies are: \n",high_movies[["title", "revenue"]].to_string(index=False))

10 most successful movies are: 
                                        title      revenue
                                      Avatar 2787965087.0
                Star Wars: The Force Awakens 2068223624.0
                                     Titanic 1845034188.0
                                The Avengers 1519557910.0
                              Jurassic World 1513528810.0
                                   Furious 7 1506249360.0
                     Avengers: Age of Ultron 1405403694.0
Harry Potter and the Deathly Hallows: Part 2 1342000000.0
                                      Frozen 1274219009.0
                        Beauty and the Beast 1262886337.0


# Create a new DataFrame ***new*** of only movies that earned more than $1 billion.

In [ ]:
new = small_df[small_df['revenue'] > 1000000000]
new

title release_date  \
1639                                         Titanic   1997-11-18   
7000   The Lord of the Rings: The Return of the King   2003-12-01   
11008     Pirates of the Caribbean: Dead Man's Chest   2006-06-20   
12481                                The Dark Knight   2008-07-16   
14551                                         Avatar   2009-12-10   
14892                            Alice in Wonderland   2010-03-03   
15348                                    Toy Story 3   2010-06-16   
17124    Pirates of the Caribbean: On Stranger Tides   2011-05-14   
17293                 Transformers: Dark of the Moon   2011-06-28   
17437   Harry Potter and the Deathly Hallows: Part 2   2011-07-07   
17818                                   The Avengers   2012-04-25   
18252                          The Dark Knight Rises   2012-07-16   
19261                                        Skyfall   2012-10-25   
19971              The Hobbit: An Unexpected Journey   2012-11-26   
20830                                     Iron Man 3   2013-04-18   
22110                                         Frozen   2013-11-27   
23617                Transformers: Age of Extinction   2014-06-25   
25084                                 Jurassic World   2015-06-09   
26555                   Star Wars: The Force Awakens   2015-12-15   
26558                        Avengers: Age of Ultron   2015-04-22   
26567                     Captain America: Civil War   2016-04-27   
28830                                      Furious 7   2015-04-01   
30700                                        Minions   2015-06-17   
36253                                       Zootopia   2016-02-11   
38176                                   Finding Dory   2016-06-16   
41489                   Rogue One: A Star Wars Story   2016-12-14   
42222                           Beauty and the Beast   2017-03-16   
43255                        The Fate of the Furious   2017-04-12   
44009                                Despicable Me 3   2017-06-15   

            budget       revenue  runtime  \
1639   200000000.0  1.845034e+09    194.0   
7000    94000000.0  1.118889e+09    201.0   
11008  200000000.0  1.065660e+09    151.0   
12481  185000000.0  1.004558e+09    152.0   
14551  237000000.0  2.787965e+09    162.0   
14892  200000000.0  1.025491e+09    108.0   
15348  200000000.0  1.066970e+09    103.0   
17124  380000000.0  1.045714e+09    136.0   
17293  195000000.0  1.123747e+09    154.0   
17437  125000000.0  1.342000e+09    130.0   
17818  220000000.0  1.519558e+09    143.0   
18252  250000000.0  1.084939e+09    165.0   
19261  200000000.0  1.108561e+09    143.0   
19971  250000000.0  1.021104e+09    169.0   
20830  200000000.0  1.215440e+09    130.0   
22110  150000000.0  1.274219e+09    102.0   
23617  210000000.0  1.091405e+09    165.0   
25084  150000000.0  1.513529e+09    124.0   
26555  245000000.0  2.068224e+09    136.0   
26558  280000000.0  1.405404e+09    141.0   
26567  250000000.0  1.153304e+09    147.0   
28830  190000000.0  1.506249e+09    137.0   
30700   74000000.0  1.156731e+09     91.0   
36253  150000000.0  1.023784e+09    108.0   
38176  200000000.0  1.028571e+09     97.0   
41489  200000000.0  1.056057e+09    133.0   
42222  160000000.0  1.262886e+09    129.0   
43255  250000000.0  1.238765e+09    136.0   
44009   80000000.0  1.020063e+09     96.0   

                                                  genres    year  
1639   [{'id': 18, 'name': 'Drama'}, {'id': 10749, 'n...  1997.0  
7000   [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...  2003.0  
11008  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...  2006.0  
12481  [{'id': 18, 'name': 'Drama'}, {'id': 28, 'name...  2008.0  
14551  [{'id': 28, 'name': 'Action'}, {'id': 12, 'nam...  2009.0  
14892  [{'id': 10751, 'name': 'Family'}, {'id': 14, '...  2010.0  
15348  [{'id': 16, 'name': 'Animation'}, {'id': 10751...  2010.0  
17124  [{'id': 12, 'name': 'Adventure'}, {'id': 28, '...  2011.0  
17293  [{'id': 28, 'na

# Create another new DataFrame new2 of only movies earned more than 1 billion, but where the budget less than 150 million.


In [ ]:
new2 = small_df[(small_df['revenue'] > 1000000000) & (small_df["budget"]< 150000000)]
new2

title release_date  \
7000   The Lord of the Rings: The Return of the King   2003-12-01   
17437   Harry Potter and the Deathly Hallows: Part 2   2011-07-07   
30700                                        Minions   2015-06-17   
44009                                Despicable Me 3   2017-06-15   

            budget       revenue  runtime  \
7000    94000000.0  1.118889e+09    201.0   
17437  125000000.0  1.342000e+09    130.0   
30700   74000000.0  1.156731e+09     91.0   
44009   80000000.0  1.020063e+09     96.0   

                                                  genres    year  
7000   [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...  2003.0  
17437  [{'id': 10751, 'name': 'Family'}, {'id': 14, '...  2011.0  
30700  [{'id': 10751, 'name': 'Family'}, {'id': 16, '...  2015.0  
44009  [{'id': 28, 'name': 'Action'}, {'id': 16, 'nam...  2017.0

# Part 2: Building a simple recommender

In [ ]:
small_df2 = movie_df.loc[:, ["title", "runtime", "vote_count", "vote_average"]]

In [ ]:
small_df2.head(10)

title  runtime  vote_count  vote_average
0                    Toy Story     81.0      5415.0           7.7
1                      Jumanji    104.0      2413.0           6.9
2             Grumpier Old Men    101.0        92.0           6.5
3            Waiting to Exhale    127.0        34.0           6.1
4  Father of the Bride Part II    106.0       173.0           5.7
5                         Heat    170.0      1886.0           7.7
6                      Sabrina    127.0       141.0           6.2
7                 Tom and Huck     97.0        45.0           5.4
8                 Sudden Death    106.0       174.0           5.5
9                    GoldenEye    130.0      1194.0           6.6

In [ ]:
# Prerequisite 1: Only considering movies that are greater than 45 minutes and less than 300 minutes in length.

new_df = small_df2[(small_df2['runtime'] > 45) & (small_df2['runtime'] < 300)]

In [ ]:
new_df['runtime'].describe() 

count    41854.000000
mean        99.622282
std         22.627991
min         46.000000
25%         88.000000
50%         96.000000
75%        108.000000
max        299.000000
Name: runtime, dtype: float64

In [ ]:
new_df.head(10)

title  runtime  vote_count  vote_average
0                    Toy Story     81.0      5415.0           7.7
1                      Jumanji    104.0      2413.0           6.9
2             Grumpier Old Men    101.0        92.0           6.5
3            Waiting to Exhale    127.0        34.0           6.1
4  Father of the Bride Part II    106.0       173.0           5.7
5                         Heat    170.0      1886.0           7.7
6                      Sabrina    127.0       141.0           6.2
7                 Tom and Huck     97.0        45.0           5.4
8                 Sudden Death    106.0       174.0           5.5
9                    GoldenEye    130.0      1194.0           6.6

In [ ]:
new_df['vote_count'].describe()

count    41854.000000
mean       118.228246
std        511.046385
min          0.000000
25%          4.000000
50%         10.000000
75%         38.000000
max      14075.000000
Name: vote_count, dtype: float64

In [ ]:
#Prerequisite 2:  
#For a movie to be considered in the rankings, it must have garnered more votes than at least 80% of the movies present in raw dataset
vote_req = new_df['vote_count'].quantile(0.8) 
movie_chart = new_df[new_df['vote_count'] > vote_req]

movie_chart

title  runtime  vote_count  \
0                                           Toy Story     81.0      5415.0   
1                                             Jumanji    104.0      2413.0   
2                                    Grumpier Old Men    101.0        92.0   
4                         Father of the Bride Part II    106.0       173.0   
5                                                Heat    170.0      1886.0   
...                                               ...      ...         ...   
45265                                   Force Majeure    118.0       255.0   
45268  Good Guys Go to Heaven, Bad Guys Go to Pattaya    100.0       153.0   
45339                                  With Open Arms     92.0        94.0   
45343                      The Visitors: Bastille Day    110.0       167.0   
45441                                     Cadet Kelly    101.0       145.0   

       vote_average  
0               7.7  
1               6.9  
2               6.5  
4               5.7  
5               7.7  
...             ...  
45265           6.8  
45268           5.3  
45339           5.2  
45343           4.0  
45441           5.2  

[8320 rows x 4 columns]

In [ ]:
m = vote_req + 1 #minimum number of votes required for the movie to be in the chart
#print(m)
c = movie_chart["vote_average"].mean() #the mean rating of all the movies in the dataset
#print(c)

# 1. Calculate the weighted rating score.
# 2. Add the score as a feature to the dataframe, by using Apply function. 

In [ ]:
def weighted_avg(x):
  wr = ((x.vote_count * x.vote_average) + (m * c))/(x.vote_count +m)
  return wr


In [ ]:
movie_chart["score"] = movie_chart.apply(weighted_avg, axis = 1 )
movie_chart

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


title  runtime  vote_count  \
0                                           Toy Story     81.0      5415.0   
1                                             Jumanji    104.0      2413.0   
2                                    Grumpier Old Men    101.0        92.0   
4                         Father of the Bride Part II    106.0       173.0   
5                                                Heat    170.0      1886.0   
...                                               ...      ...         ...   
45265                                   Force Majeure    118.0       255.0   
45268  Good Guys Go to Heaven, Bad Guys Go to Pattaya    100.0       153.0   
45339                                  With Open Arms     92.0        94.0   
45343                      The Visitors: Bastille Day    110.0       167.0   
45441                                     Cadet Kelly    101.0       145.0   

       vote_average     score  
0               7.7  7.686082  
1               6.9  6.887627  
2               6.5  6.447914  
4               5.7  5.864518  
5               7.7  7.660802  
...             ...       ...  
45265           6.8  6.720318  
45268           5.3  5.588758  
45339           5.2  5.639333  
45343           4.0  4.601514  
45441           5.2  5.528412  

[8320 rows x 5 columns]

#3. Sort the dataframe on the basis of the score just computed.
#4. Output of the list of top movies. 

In [ ]:
sorted_df = movie_chart.sort_values('score', ascending=False).head(10)
sorted_df.loc[:,["title","runtime","vote_average","score"]]

title  runtime  vote_average     score
10309  Dilwale Dulhania Le Jayenge    190.0           9.1  8.882784
314       The Shawshank Redemption    142.0           8.5  8.485531
834                  The Godfather    175.0           8.5  8.479977
40251                   Your Name.    106.0           8.5  8.387985
12481              The Dark Knight    152.0           8.3  8.291047
2843                    Fight Club    139.0           8.3  8.288664
292                   Pulp Fiction    154.0           8.3  8.287354
522               Schindler's List    195.0           8.3  8.275437
23673                     Whiplash    105.0           8.3  8.275105
5481                 Spirited Away    125.0           8.3  8.272581

# Part 3:  Implement a user-based collaborative filtering system

#1. Using the data in ratings_small.csv, construct the user rating matrix (where the rows represent the users, and the columns the movies).

In [ ]:
df_ratings = pd.read_csv("ratings_small.csv")

In [ ]:
df_ratings.head(10)

userId  movieId  rating   timestamp
0       1       31     2.5  1260759144
1       1     1029     3.0  1260759179
2       1     1061     3.0  1260759182
3       1     1129     2.0  1260759185
4       1     1172     4.0  1260759205
5       1     1263     2.0  1260759151
6       1     1287     2.0  1260759187
7       1     1293     2.0  1260759148
8       1     1339     3.5  1260759125
9       1     1343     2.0  1260759131

In [ ]:
user_ratingmatrix = df_ratings.pivot_table(index = "userId", columns = "movieId", values = "rating")

In [ ]:
user_ratingmatrix #matrix showing the user preferences 

movieId  1       2       3       4       5       6       7       8       \
userId                                                                    
1           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
2           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
3           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
4           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
5           NaN     NaN     4.0     NaN     NaN     NaN     NaN     NaN   
...         ...     ...     ...     ...     ...     ...     ...     ...   
667         NaN     NaN     NaN     NaN     NaN     4.0     NaN     NaN   
668         NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
669         NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
670         4.0     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
671         5.0     NaN     NaN     NaN     NaN     NaN     NaN     NaN   

movieId  9       10      ...  161084  161155  161594  161830  161918  161944  \
userId                   ...                                                   
1           NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN     NaN   
2           NaN     4.0  ...     NaN     NaN     NaN     NaN     NaN     NaN   
3           NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN     NaN   
4           NaN     4.0  ...     NaN     NaN     NaN     NaN     NaN     NaN   
5           NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN     NaN   
...         ...     ...  ...     ...     ...     ...     ...     ...     ...   
667         NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN     NaN   
668         NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN     NaN   
669         NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN     NaN   
670         NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN     NaN   
671         NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN     NaN   

movieId  162376  162542  162672  163949  
userId                                   
1           NaN     NaN     NaN     NaN  
2           NaN     NaN     NaN     NaN  
3           NaN     NaN     NaN     NaN  
4           NaN     NaN     NaN     NaN  
5           NaN     NaN     NaN     NaN  
...         ...     ...     ...     ...  
667         NaN     NaN     NaN     NaN  
668         NaN     NaN     NaN     NaN  
669         NaN     NaN     NaN     NaN  
670         NaN     NaN     NaN     NaN  
671         NaN     NaN     NaN     NaN  

[671 rows x 9066 columns]

# 2. This matrix (user_ratingmatrix) will be sparse (lot of unknown values), explain why. Propose a way to fill in the missing values and justify it. Implement it to have a matrix without missing values.

***Answer:*** The matrix rows represents the users and the columns represents the movies. If a particular movie has been rated by a particular user, only then the value appers in that particular position in the matrix. Since there are 9066 columns i.e, there are 9066 movies, for a user it's kinda unlikely that all movies to be rated. Additionally it's unlikely to expect that a particular movie to be rated by all users. That's why there are a lot of unknown values in the sparse matrix.

We can think of filling up the unknown values by 0. Because we don't really know the values that a user could rate for particular movie. If we take the approach to fill up the known values of the movies by the average of all ratings given by the user or by the average rating of a movie that can be misleading and biased. 

# 3. Calculate the cosine similarity for each pair of users, in order to obtain a user-user similarity matrix.

In [ ]:
from sklearn.metrics.pairwise import cosine_similarity
cosinesim_user = cosine_similarity(user_ratingmatrix.fillna(0))
np.fill_diagonal(cosinesim_user, 0) # the diagonals values signifies the same user for which the cosine sim. value is 1, and it's not necessary for the calculation. Hence we make it 0. 
df_cosinesim_user = pd.DataFrame(cosinesim_user, index = user_ratingmatrix.index, columns=user_ratingmatrix.index)

In [ ]:
df_cosinesim_user #user-user similarity matrix

userId       1         2         3         4         5         6         7    \
userId                                                                         
1       0.000000  0.000000  0.000000  0.074482  0.016818  0.000000  0.083884   
2       0.000000  0.000000  0.124295  0.118821  0.103646  0.000000  0.212985   
3       0.000000  0.124295  0.000000  0.081640  0.151531  0.060691  0.154714   
4       0.074482  0.118821  0.081640  0.000000  0.130649  0.079648  0.319745   
5       0.016818  0.103646  0.151531  0.130649  0.000000  0.063796  0.095888   
...          ...       ...       ...       ...       ...       ...       ...   
667     0.000000  0.425462  0.124562  0.088735  0.058252  0.000000  0.232051   
668     0.000000  0.084646  0.124911  0.068483  0.042926  0.019563  0.058773   
669     0.062917  0.024140  0.080984  0.104309  0.038358  0.024583  0.073151   
670     0.000000  0.170595  0.136606  0.054512  0.062642  0.019465  0.096240   
671     0.017466  0.113175  0.170193  0.211609  0.225086  0.087705  0.268672   

userId       8         9         10   ...       662       663       664  \
userId                                ...                                 
1       0.000000  0.012843  0.000000  ...  0.000000  0.000000  0.014474   
2       0.113190  0.113333  0.043213  ...  0.477306  0.063202  0.077745   
3       0.249781  0.134475  0.114672  ...  0.161205  0.064198  0.176134   
4       0.191013  0.030417  0.137186  ...  0.114319  0.047228  0.136579   
5       0.165712  0.086616  0.032370  ...  0.191029  0.021142  0.146173   
...          ...       ...       ...  ...       ...       ...       ...   
667     0.069005  0.066412  0.032653  ...  0.342283  0.050754  0.074080   
668     0.112366  0.194493  0.098561  ...  0.074089  0.059010  0.093021   
669     0.055143  0.029291  0.060549  ...  0.015960  0.025953  0.077927   
670     0.247687  0.384429  0.158650  ...  0.183662  0.122126  0.123407   
671     0.406414  0.168497  0.189703  ...  0.096704  0.070749  0.219728   

userId       665       666       667       668       669       670       671  
userId                                                                        
1       0.043719  0.000000  0.000000  0.000000  0.062917  0.000000  0.017466  
2       0.164162  0.466281  0.425462  0.084646  0.024140  0.170595  0.113175  
3       0.158357  0.177098  0.124562  0.124911  0.080984  0.136606  0.170193  
4       0.254030  0.121905  0.088735  0.068483  0.104309  0.054512  0.211609  
5       0.224245  0.139721  0.058252  0.042926  0.038358  0.062642  0.225086  
...          ...       ...       ...       ...       ...       ...       ...  
667     0.124214  0.351207  0.000000  0.091597  0.018416  0.153111  0.127995  
668     0.082525  0.114487  0.091597  0.000000  0.000000  0.178017  0.135387  
669     0.101707  0.028773  0.018416  0.000000  0.000000  0.042609  0.085202  
670     0.143380  0.159479  0.153111  0.178017  0.042609  0.000000  0.228677  
671     0.239318  0.127204  0.127995  0.135387  0.085202  0.228677  0.000000  

[671 rows x 671 columns]

In [ ]:
h = df_cosinesim_user.loc[1,:].sort_values(axis = 0, ascending=False)
print(h.head(10)) #Top 10 users who are similar to the user with userId: 1. 

userId
325    0.371852
634    0.194093
341    0.162819
310    0.157524
207    0.152746
35     0.130585
195    0.122647
485    0.114021
130    0.112817
229    0.112575
Name: 1, dtype: float64


# 4.  Implement a function to return the n nearest neighbors of a given user in the sense of the calculated similarity.

In [ ]:
def n_neighbor(k, n):       #k is the user for whom we are calculating the n number of nearest neighbors.
  neighbors = df_cosinesim_user.loc[k,:].sort_values(axis = 0, ascending=False)
  return neighbors.head(n)

call1 = n_neighbor(1,5)

print("5 nearest neighbors of the user_id= 1:",list(call1.index))


5 nearest neighbors of the user_id= 1: [325, 634, 341, 310, 207]


# 5. Using the preceding function, implement a function that recommends k movies—at most—for a given user. Explain how this function works.


In [ ]:
def score_function(x): #It is the score function to calculate the weighted movie scores of the movies reviewed by similar user
  s = []
  d = []
  score = []
  l = np.unique(list(x.movieId))
  for i in l: 
    s = list(x[x["movieId"]==i]["movie_id_cos_mul"]) #obtiaining all the (cosinesimilarity * movie rating) of a particular movie
    sum_s = sum(s)
    d = list(x[x["movieId"]==i]["cosine_value"]) #obtiaining the cosine similarity of all users who has rated that particular movie
    sum_d = sum(d)
    if sum(d) > 0.0 and sum(s) > 0.0:
      score.append(sum_s/sum_d) #adding the weighted score to a list
    else:
      score.append(0)
  return score

In [ ]:
def movie_recommender(n,k,n_nearest = 8): #for userId n we are calculating the  k movie recommendations.

  similar_userton = n_neighbor(n,n_nearest) # By default we will be calculating the 8 neartest neighbor to n.
  dict_similar_userton = similar_userton.to_dict() #Converting similar_userton  to a dictionary 
  recommendar_group = df_ratings[df_ratings['userId'].isin(list(similar_userton.index))] #data frame containing the movie rating of sim. users.
  
  user_n = df_ratings[df_ratings["userId"].isin([n])] #movie ratings of the user with the userId n
  a = list(user_n["movieId"]) # making a list that has already been rated by the userId n

  #filtered_recommendar_group is the filtered dataframe containing the movies that aren't yet watched by userId n
  filtered_recommendar_group = recommendar_group[~recommendar_group["movieId"].isin(a)]

  #Mapping the cosine values of the users corresponding to the userIds 
  filtered_recommendar_group["cosine_value"] = filtered_recommendar_group["userId"].map(dict_similar_userton)

  # Multiplying the cosine similarity of a particular user with the all movies that has been watched by the user
  filtered_recommendar_group["movie_id_cos_mul"] = filtered_recommendar_group["cosine_value"] * filtered_recommendar_group["rating"]
  #sorting the filtered_recommendar_group accoring to the movie Id's 
  sorted_filtered_rg = filtered_recommendar_group.sort_values('movieId')
  smallsorted_filtered = sorted_filtered_rg.loc[:, ["movieId", "cosine_value", "movie_id_cos_mul"]]

  #Calling the score_function that was defined before movie_recommender() to calculate the weighted scores of the movies that are going to be recommeded on the basis of the user similarity.
  score_list = score_function(smallsorted_filtered)
  unique_movieIds = np.unique(list(smallsorted_filtered.movieId))

  #Creating a dataframe containing all the movies along with their weighted scores 
  rec_movies = pd.DataFrame(list(zip(unique_movieIds, score_list)), columns = ["movieId", "score"])

  #Creating a dataframe containg the k number of movieIds with the heighest scores 
  sorted_rec_movies = rec_movies.sort_values("score", ascending=False).head(k)
  return sorted_rec_movies



c = movie_recommender(2,10)
print("Reommended movie list for userId 2 is: \n ", c)

Reommended movie list for userId 2 is: 
      movieId     score
20      111  5.000000
75      477  5.000000
3         6  5.000000
39      230  5.000000
5        16  5.000000
81      594  5.000000
55      318  4.829193
59      337  4.345258
85      608  4.248665
80      555  4.012255


/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:14: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:17: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy


# 6. Propose a procedure to find which value of n is appropriate to determine the recommendations for this dataset.

***Answer:*** In the user based collaborative recommender system, the n nearest neighbor to a particular user plays an important role in determining the content recommendation. To determine the value of n we should consider a threshold value for the cosine similarity between users. For example we can only consider the users who's consine similarity with a particular user is more than 90% of the users'cosine similarity.